src/rpc/cli.cpp

```diff
@@ -114,57 +114,75 @@ void cli::run()
 /****
  * @brief loop through list of commands, attempting to find a match
  * @param token what the user typed
- * @param match
- * @returns the full name of the command or NULL if no matches found
+ * @param match sets to 1 if only 1 match was found
+ * @returns the remaining letters of the name of the command or NULL if 1 match not found
  */
 static char *my_rl_complete(char *token, int *match)
 {
-    int matchlen = 0;
-    int count = 0;
-    const char* method_name;
-
-    auto& cmd = cli_commands();
-    int partlen = strlen (token); /* Part of token */
-    for (auto it : cmd) {
-    	if (!strncmp (it.c_str(), token, partlen)) {
-    		method_name = it.c_str();
-    		matchlen = partlen;
-    		count ++;
-    	}
-    }
-
-    if (count == 1) {
-    	*match = 1;
-    	return strdup (method_name + matchlen);
-    }
-
-    return NULL;
+   bool have_one = false;
+   std::string method_name;
+
+   auto& cmd = cli_commands();
+   int partlen = strlen (token); /* Part of token */
+
+   for (const std::string it : cmd)
+   {
+      if (it.compare(0, partlen, token) == 0)
+      {
+         if (have_one) {
+            // we can only have 1, but we found a second
+            return NULL;
+         }
+         else
+         {
+            method_name = it;
+            have_one = true;
+         }
+      }
+   }
+
+   if (have_one)
+   {
+      *match = 1;
+      method_name += " ";
+      return strdup (method_name.c_str() + partlen);
+   }
+
+   return NULL;
 }
 
 /***
  * @brief return an array of matching commands
  * @param token the incoming text
- * @param av the resultant array of possible matches
+ * @param array the resultant array of possible matches
  * @returns the number of matches
  */
-static int cli_completion(char *token, char ***av)
+static int cli_completion(char *token, char ***array)
 {
-    int num, total = 0;
-    char **copy;
-
-    auto& cmd = cli_commands();
-    num = cmd.size();
-
-    copy = (char **) malloc (num * sizeof(char *));
-    for (auto it : cmd) {
-    	if (!strncmp (it.c_str(), token, strlen (token))) {
-    		copy[total] = strdup ( it.c_str() );
-    		total ++;
-    	}
-    }
-    *av = copy;
-
-    return total;
+   auto& cmd = cli_commands();
+   int num_commands = cmd.size();
+
+   char **copy = (char **) malloc (num_commands * sizeof(char *));
+   if (copy == NULL)
+   {
+      // possible out of memory
+      return 0;
+   }
```

Comment:
I want to make sure the memory allocated here and those allocated by calling `strdup(...)` elsewhere will be correctly freed, to avoid memory leak.

Reply:
Ok, understood. The memory allocated for the list of method names generated by cli_completion is freed internally by editline ( see https://github.com/troglobit/editline/blob/5be965deec0527b48b393b7aca9cb2a76aec2a99/src/editline.c#L1533 ). 

The memory allocated to complete the current method name in my_rl_complete is freed by us ( https://github.com/bitshares/bitshares-fc/blob/b8e5640238aac6b29e1829dc07f47dbc4873d9da/src/rpc/cli.cpp#L204 ).

Reply:
The free() is never called if add_history() throws.
This is why malloc/free should generally be avoided, for example by wrapping pointers in an object (like shared_ptr).

Reply:
IMHO, at this stage, just surround `add_history()` with `try` and `catch`, or use `shared_ptr`, to make sure we'll free the memory. Then we merge the PR and do other things including the partially completion feature later. All realine-related features in CLI are not critical, what we need now is a working replacement of readline, don't need to make it perfect immediately.

Reply:
Used try/catch.